src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+            _logger = loggerFactory?.CreateLogger<RazorHoverEndpoint>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Override to provide fallback behavior for when single server is not supported
+        /// </summary>
+        protected abstract Task<TResponse?> HandleWithoutSingleServerAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Override to provide behavior for razor contexts. Since the c#/html servers won't
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="EndpointName"/>
+        /// </summary>
+        protected abstract TDelegatedParams CreateDelegatedParams(TRequest request, DocumentContext documentContext, Projection projection);
+
+        /// <summary>
+        /// The name of the endpoint to delegate to, from <see cref="LanguageServerConstants"/>
+        /// </summary>
+        protected abstract string EndpointName { get; }
+
+        /// <summary>
+        /// If the response needs to be remapped for any reasons, override to handle the remapping logic
+        /// </summary>
+        protected virtual Task<TResponse> RemapResponseAsync(TResponse delegatedResponse, RazorCodeDocument codeDocument, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// Extracts information about the request and then calls <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>.
+        /// Children can override to provide any more custom logic needed before doing a normal delegated response pass with <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
```

Comment:
The unsupported check historically was to ensure in VSCode scenarios we didn't do any extra work in .NET Framework apps for Razor. Part of me feels like we should just straight up remove all of the unsupported bits and experience the fallout from the VSCode side of the world / determine a better way to handle that situation

Reply:
I'll just remove. Be bold I guess

Reply:
💯 love it